### PR TITLE
Little refactoring of `matutils.corpus2csc`

### PR DESCRIPTION
**Pull request by horpto**
- ease check on fast and slow version
- make more relaxed for num_doc and num_terms arguments

TODO: 

- [x] tests
- [x] check on empty doc
- [x] benchmark

bench:
[corpus2csc.txt](https://github.com/RaRe-Technologies/gensim/files/2815531/corpus2csc.txt)

bench results:
```
CLASSIC, fast path
1.9 s ± 5.61 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
CLASSIC, slow path
2.63 s ± 19.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
NEW, fast path
1.42 s ± 8.86 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
NEW, slow path
2.45 s ± 6.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
NEW 2, fast path
1.41 s ± 15.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
NEW 2, slow path
1.43 s ± 10.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
NEW 3, fast path
1.94 s ± 14.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
NEW 3, slow path
1.95 s ± 7.24 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

### CODE REVIEW

File: gensim/matutils.py [hunk 1]
```diff
@@ -119,51 +119,42 @@ def corpus2csc(corpus, num_terms=None, dtype=np.float64, num_docs=None, num_nnz=
...
+        num_terms = indices.max() + 1 if num_nnz > 0 else 0
+
+    if num_docs is None:
         num_docs = len(indptr) - 1
-        # now num_docs, num_terms and num_nnz contain the correct values
-        data = np.asarray(data, dtype=dtype)
-        indices = np.asarray(indices)
-        result = scipy.sparse.csc_matrix((data, indices, indptr), shape=(num_terms, num_docs), dtype=dtype)
+    else:
+        assert num_docs == len(indptr) - 1, "mismatch between supplied and computed number of documents"
```
Comment (mpenkov) [inline gensim/matutils.py]: I think it's helpful to include the actual numbers in the assert message;

```suggestion
        assert num_docs == len(indptr) - 1, "supplied docs (%d) != computed docs (%d)" % (num_docs, len(indptr) - 1)
```

File: gensim/matutils.py [hunk 2]
```diff
@@ -119,51 +119,42 @@ def corpus2csc(corpus, num_terms=None, dtype=np.float64, num_docs=None, num_nnz=
...
+        if printprogress and docno % printprogress == 0:
+            logger.info("PROGRESS: at document #%i/%i", docno, num_docs if num_docs is not None else "-")
+        posnext = posnow + len(doc)
+        # don't check like `if doc:` on the case if doc has numpy.array dtype
+        if posnext > posnow:
+            # transpose doc and set column indices and data
+            indices[posnow: posnext], data[posnow: posnext] = zip(*doc)
+        indptr.append(posnext)
+        posnow = posnext
+    assert posnow == num_nnz, "mismatch between supplied and computed number of non-zeros"
```
Comment (mpenkov) [inline gensim/matutils.py]:
```suggestion
    assert posnow == num_nnz, "mismatch between supplied and computed number of non-zeros (%d != %d)" % (posnow, num_nnz)
```

File: gensim/matutils.py [hunk 3]
```diff
@@ -119,51 +119,42 @@ def corpus2csc(corpus, num_terms=None, dtype=np.float64, num_docs=None, num_nnz=
...
-            if printprogress and docno % printprogress == 0:
-                logger.info("PROGRESS: at document #%i", docno)
-            indices.extend(feature_id for feature_id, _ in doc)
-            data.extend(feature_weight for _, feature_weight in doc)
-            num_nnz += len(doc)
-            indptr.append(num_nnz)
-        if num_terms is None:
-            num_terms = max(indices) + 1 if indices else 0
+    if num_nnz is None:
+        num_nnz = sum(len(doc) for doc in corpus)
```
Comment (piskvorky) [inline gensim/matutils.py]: `corpus` may be (and often is) a generator. We cannot afford any extra passes like this. Make sure the refactoring keeps it single-pass.

**Comment by piskvorky** [comment]
@horpto The extra tests are nice, thanks.

What is `NEW` / `NEW 2` / `NEW 3` in your PR description? Why are these numbers different?

Also, please include your motivation in the PR description – did users complain about something? Did you find problems in your own workflows? What were they? Why these set of changes?

We require each PR to be clearly motivated in order to be accepted. Going forward with Gensim, new functionality without a clear motivation will not be accepted and existing functionality without a clear motivation will be removed from Gensim (deprecated and then ripped out). CC @mpenkov .

**Comment by mpenkov** [comment]
@horpto What is the status of this PR? Are you able to finish it? If not, it may be best to close it.

**Pull request closed without merging**